### PR TITLE
mon/OSDMonitor: batch OSDs nodown/noout support

### PR DESCRIPTION
**Pull request by xiexingguo**
This patch allow us to add batch OSDs(e.g., from a specific host which is currently in maintenance)
into a specific nodown/noout list, which can not be automatically marked down/out
and hence can be eliminated from data migration.

This has the same effect with the global nodown/noout flag but is more fine-grained.

Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>

### CODE REVIEW

File: src/osd/OSDMap.h [hunk 1]
```diff
@@ -174,6 +174,10 @@ class OSDMap {
     float new_full_ratio = -1;
 
     int8_t new_require_min_compat_client = -1;
+    vector<int32_t> new_noout_osds;
```
Comment (branch-predictor) [inline src/osd/OSDMap.h]: Use a bit vector/array for that, so each osd can be mapped to single bit and located in constant time *and* take less memory at the same time.

File: src/osd/OSDMap.cc [hunk 2]
```diff
@@ -505,6 +505,10 @@ void OSDMap::Incremental::encode(bufferlist& bl, uint64_t features) const
     if (target_v >= 6) {
       ::encode(new_require_min_compat_client, bl);
       ::encode(new_require_osd_release, bl);
+      ::encode(new_noout_osds, bl);
```
Comment (branch-predictor) [inline src/osd/OSDMap.cc]: Shouldn't that go into a separate version (and separate target_v check)?

File: src/osd/OSDMap.h [hunk 3]
```diff
@@ -174,6 +174,10 @@ class OSDMap {
...
+    vector<int32_t> old_noout_osds;
+    vector<int32_t> new_nodown_osds;
+    vector<int32_t> old_nodown_osds;
```
Comment (liewegas) [inline src/osd/OSDMap.h]: I think the way to do this is actually with the osd_state bits... except that we're out of bits.  I can do a quick PR to expand it to a uint32_t, though, with the switch to luminous.  Then you can base this PR on top of that.

**Comment by liewegas** [comment]
See https://github.com/ceph/ceph/pull/15390

**Comment by xiexingguo** [comment]
> See #15390

Done rebasing on that. @liewegas

File: src/mon/MonCommands.h [hunk 4]
```diff
@@ -650,6 +650,22 @@ COMMAND("osd in " \
...
+        "name=ids,type=CephString,n=N", \
+        "mark osd(s) <id> [<id>...] as noout", \
+        "osd", "rw", "cli,rest")
+COMMAND("osd add-nodown-osds " \
+        "name=ids,type=CephString,n=N", \
+        "mark osd(s) <id> [<id>...] as nodown", \
+        "osd", "rw", "cli,rest")
+COMMAND("osd rm-noout-osds " \
+        "name=ids,type=CephString,n=N", \
+        "allow osd(s) <id> [<id>...] to be marked out(if they are currently marked as noout)", \
```
Comment (liewegas) [inline src/mon/MonCommands.h]: s/out(/out (/
same below

File: src/mon/OSDMonitor.cc [hunk 5]
```diff
@@ -8417,6 +8458,204 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
...
+             prefix == "osd add-nodown-osds") {
+
+    bool noout = prefix == "osd add-noout-osds";
+    bool any = false;
+
+    vector<string> idvec;
+    cmd_getval(g_ceph_context, cmdmap, "ids", idvec);
+    for (unsigned j = 0; j < idvec.size(); j++) {
+      long osd_start = -1;
+      long osd_end = -1;
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: nit: usually use "end" in exclusive sense, like [begin, end).. for inclusive maybe [first,last].

**Comment by liewegas** [comment]
@jdurgin thoughts on this?  I'm not wild about rm-noout-osds and add-noout-osds (maybe drop -osds suffix?).  Also unsure about the range syntax; the caller can do that with $(seq a b).

**Comment by liewegas** [comment]
also i suspect a more useful unit of specifying osds is via the crush hierarchy (so that you mark a host,rack,etc as noout or whatever).  That suggests a generic syntax for specifying osds that is used for all of the related commands (osd down, osd out, etc.).

**Comment by liewegas** [comment]
...or a 'ceph osd ls-tree host4` that spits out the list of osds so that you do the substitution in the shell...

**Comment by jdurgin** [comment]
I'd rather leave out the range syntax as well. I like the 'ceph osd ls-tree' idea, and just accepting a list of osd numbers here - it'd be useful for other osd-related commands too

**Comment by xiexingguo** [comment]
> I'd rather leave out the range syntax as well.

Actually, it's compatible with the old fashion(`ceph osd add-noout-osds 0 1 2` still works) and is more flexible, e.g., allow you do it in a mixture of both, e.g.:

`ceph osd add-noout-osds 0 1~2`

The benefit is that for large clusters, we do not want to transport a long list osds, which is annoying for Calamari(or anything like that).

**Comment by xiexingguo** [comment]
> I like the 'ceph osd ls-tree' idea

Will give a try.

File: src/mon/OSDMonitor.cc (hunk 5 above)
Comment (xiexingguo) [inline src/mon/OSDMonitor.cc]: Updated.

**Comment by xiexingguo** [comment]
changelog:
- add-noout-osds -> add-noout
- add "ceph osd ls-tree " cmd

Local test result:

```
    ~# ceph osd tree
    ID WEIGHT  TYPE NAME                                        UP/DOWN REWEIGHT PRIMARY-AFFINITY
    -5 3.00000 root foo
    -4 2.00000     rack foo-rack-1
    -3 1.00000         host foo-host-192-8-8-91
     1 1.00000             osd.1                                     up  1.00000          1.00000
    -6 1.00000         host foo-host-192-8-8-92
     2 1.00000             osd.2                                     up  1.00000          1.00000
    -8 1.00000     rack foo-rack-2
    -7 1.00000         host foo-host-192-8-8-93
     0 1.00000             osd.0                                     up  1.00000          1.00000
    -1       0 root default
    -2       0     host gitbuilder-ceph-rpm-centos7-amd64-basic
    
    ~# ceph osd ls-tree foo
    0
    1
    2
    
    ~# ceph osd ls-tree foo-rack-1
    1
    2
    
    ~# ceph osd ls-tree foo-rack-2
    0
    
    ~# ceph osd ls-tree foo-host-192-8-8-92
    2
```

@liewegas  Can you look again?

**Comment by xiexingguo** [comment]
retest this please

File: src/mon/OSDMonitor.cc [hunk 6]
```diff
@@ -8417,6 +8458,204 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
...
+              any = true;
+            }
+          } else {
+            if (!osdmap.is_nodown(p.first) &&
+                (p.second & CEPH_OSD_NODOWN) != 0) {
+              p.second &= ~CEPH_OSD_NODOWN;
+              any = true;
+            }
+          }
+        }
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: I wonder if we'd be better served with new helper functions like _pending_osd_state_{set,clear}(int osd, unsigned bit) that hide this ugliness.  I see that 'osd down' and most of the other places that touch new_state do bother to check pending_inc and can clobber a racing update.  That'll simplify this quite a bit.  (Can leave fixing the other would-be users as a follow-up)

File: src/mon/OSDMonitor.cc [hunk 7]
```diff
@@ -8417,6 +8458,204 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
...
+          auto first = idvec[j].substr(0, delimeter_pos);
+          auto last = idvec[j].substr(delimeter_pos + 1,
+                                      idvec[j].length() - delimeter_pos - 1);
+          osd_first = parse_osd_id(first.c_str(), &ss);
+          osd_last = parse_osd_id(last.c_str(), &ss);
+        } else {
+          // try single osd way
+          osd_first = parse_osd_id(idvec[j].c_str(), &ss);
+          osd_last = osd_first;
+        }
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: I'm still not wild about the range syntax here.  The only time i can see thsi being useful is when you're marking the entire cluster down  or noout or whatever.  In practice, osd ids won't be sequential ranges on hosts, so I'm not sure it really helps anyone.

**Comment by liewegas** [comment]
ls-tree looks great!

Thanks @xiexingguo and sorry for all the bike-shedding!

**Comment by xiexingguo** [comment]
@liewegas  Sorry for the late response(I was travelling for Ceph Day Beijing last weekend).

I think most of the above issues have been addressed now. Can you recheck?

**Comment by xiexingguo** [comment]
retest this please

**Comment by yuriw** [comment]
100% tests passed, 0 tests failed out of 173

http://pulpito.ceph.com/yuriw-2017-06-13_19:46:02-rados-wip-yuri-testing2_2017_7_14-distro-basic-smithi/

http://pulpito.ceph.com/yuriw-2017-06-14_00:42:11-rados-wip-yuri-testing2_2017_7_14-distro-basic-smithi/

**Pull request merged**